src/rgw/rgw_rados.cc

```diff
@@ -9158,38 +9162,53 @@ int RGWRados::cls_bucket_list_unordered(RGWBucketInfo& bucket_info,
     " start " << start.name << "[" << start.instance <<
     "] num_entries " << num_entries << dendl;
 
+  static MultipartMetaFilter multipart_meta_filter;
+
   *is_truncated = false;
   librados::IoCtx index_ctx;
 
-  rgw_obj_index_key my_start = start;
-
   map<int, string> oids;
   int r = open_bucket_index(bucket_info, index_ctx, oids, shard_id);
   if (r < 0)
     return r;
   const uint32_t num_shards = oids.size();
 
+  rgw_obj_index_key marker = start;
   uint32_t current_shard;
   if (shard_id >= 0) {
     current_shard = shard_id;
-  } else if (my_start.empty()) {
+  } else if (start.empty()) {
     current_shard = 0u;
   } else {
-    current_shard =
-      rgw_bucket_shard_index(my_start.name, num_shards);
+    std::string oid;
+    if(! multipart_meta_filter.filter(start.name, oid)) {
+      // if multipart_meta_filter fails, must be "regular" and just
+      // use name
+      oid = start.name;
+    }
+    rgw_obj_key obj_key;
+    ceph_assert(rgw_obj_key::parse_raw_oid(oid, &obj_key));
+    if (obj_key.name.empty()) {
+      // if the name is now empty, then the start marker that came in
+      // only had a namespace and nothing else and we need to start
+      // from the beginning
+      current_shard = 0u;
+    } else {
+      current_shard = rgw_bucket_shard_index(obj_key.name, num_shards);
+    }
   }
 
   uint32_t count = 0u;
   map<string, bufferlist> updates;
-  std::string last_added_entry;
+  rgw_obj_index_key last_added_entry;
   while (count <= num_entries &&
 	 ((shard_id >= 0 && current_shard == uint32_t(shard_id)) ||
 	  current_shard < num_shards)) {
     // key   - oid (for different shards if there is any)
     // value - list result for the corresponding oid (shard), it is filled by
     //         the AIO callback
     map<int, struct rgw_cls_list_ret> list_results;
```

Comment:
hmm;  if this is dense, then should be a vector

Reply:
At least in this case, the code is reading one bucket index shard at a time, so in fact it's a map with just one entry. The oids variable passed into the call (the CLSRGWIssueBucketList functor) represents the requested bucket index shards, and is  a mapping from bucket index shard numbers to oid strings. So this call is structured around requesting listings of subsets of bucket index shards.

Perhaps in cases where the request goes to *all* bucket index shards there should be a variation of the call that uses vectors.